### PR TITLE
feat(helper): propose untrusted-labeler logins via a full-history sweep

### LINKED ISSUE #2670: feat(helper): propose untrusted-labeler logins via a full-history sweep
Parent roadmap: #2668

## Background

`docs/customization.md`'s Reserved-label guard recipe already documents a
full-history sweep technique for building the untrusted-labeler login
list (#1928): a paginated `GET /repos/{owner}/{repo}/issues/events` read,
filtered to `event == "labeled"` entries whose actor `type == "Bot"`,
covering issues and pull requests in one endpoint. Today an adopter must
run this by hand — read the prose, construct the `gh api --paginate`
call, and filter the JSON themselves. This track automates the sweep as a
standalone, read-only reporting helper.

## Proposed change

Add a new helper, following this repository's TypeScript-source
convention (`docs/typescript-sources.md`: `.mts` source, generated
`.mjs`, `bin/` shim):

- `src/scripts/idd-suggest-untrusted-labelers.mts`, generating
  `scripts/idd-suggest-untrusted-labelers.mjs` and a matching `bin/` shim.
- Paginates `GET /repos/{owner}/{repo}/issues/events` to completion,
  filters to `event == "labeled"` entries whose actor `type == "Bot"`,
  and deduplicates by login.
- Prints each distinct bot login found, together with a count of
  `labeled` events attributed to it, so the operator can judge how
  frequently each one actually labels.
- Performs **no write operations** of any kind: it never edits
  `.github/idd/config.json` and never mutates GitHub state (no label
  removal, no comments) — this track ships a reporting tool only. Adding
  a login to `labels.untrustedLabelerLogins` stays a manual, adopter-owned
  edit.

## Acceptance criteria

- The helper paginates `GET /repos/{owner}/{repo}/issues/events` to
  completion and filters to `event == "labeled"` entries with
  `actor.type == "Bot"`.
- Output lists each distinct bot login with its labeling-event count.
- The helper makes no write calls of any kind (config file or GitHub
  API) — verified by a unit test asserting no mutating call is made.
- A unit test exercises pagination and filtering against fixture event
  pages (mocked; no live network call in tests).
- Follows this repository's TypeScript-source convention: `.mts` source,
  generated `.mjs`, and a `bin/` shim.
- `pnpm run typecheck`, `pnpm run build`, and `node --test tests/*.test.mts`
  pass.
- `pnpm run lint` passes.

## Candidate files

- `src/scripts/idd-suggest-untrusted-labelers.mts`
- `bin/idd-suggest-untrusted-labelers.mjs`
- `tests/idd-suggest-untrusted-labelers.test.mts`
- `docs/idd-helper-scripts.md`

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

_Effort: M -- author-estimated size; a soft autopilot
selection tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
# Summary

Adds `idd-suggest-untrusted-labelers`, a new read-only helper that
automates `docs/customization.md`'s Reserved-label guard recipe's
"full-history sweep" technique: it paginates `GET
/repos/{owner}/{repo}/issues/events` to completion, filters to `event
== "labeled"` entries whose actor `type == "Bot"`, deduplicates by
login, and prints each distinct bot login with a count of `labeled`
events attributed to it — so an adopter no longer has to hand-construct
the `gh api --paginate` call and filter the JSON themselves before
building `strip-untrusted-labels.yml`'s `<labeler-bot-login-N>` list.

Follows this repository's TypeScript-source convention (`.mts` source →
generated `.mjs` → `bin/` shim): `src/scripts/idd-suggest-untrusted-labelers.mts`
generates `scripts/idd-suggest-untrusted-labelers.mjs` and
`bin/idd-suggest-untrusted-labelers.mjs`, registered in
`HELPER_COMMANDS` and `package.json`'s `bin` map.

**Performs no write operation of any kind** — no `.github/idd/config.json`
edit, no GitHub mutation. It pages manually (`page=1,2,...&per_page=100`,
one `gh api` GET per page with a size-reducing `--jq` projection) rather
than a single `gh api --paginate` subprocess call, because the
repository-level `issues/events` endpoint embeds the full parent `issue`
object in every event and this repository's synchronous `gh`-exec path
has no `maxBuffer` override — a single buffered call could exceed Node's
default 1 MiB cap on a large repository's full history.

Test coverage (`tests/idd-suggest-untrusted-labelers.test.mts`, 13 cases):
fixture-driven tests for the pure aggregation function; dependency-injected
unit tests for the pagination loop; and three stub-`gh`-on-`PATH`
CLI-subprocess tests (mirroring `tests/gh-pagination-parsing-smoke.test.mts`'s
existing pattern) that run the compiled `.mjs` against an exact-argv
response table — since the table registers only the expected read `GET`
calls (covering both the explicit `--owner`/`--repo` path and the
`gh repo view` auto-detect path), any other or mutating `gh` invocation
fails the stub loudly, which is the "no mutating call" proof the issue's
acceptance criteria call for.

A C1 critique pass (both a fresh-subagent review and self-review) on
the initial diff caught two doc/comment inaccuracies (a fabricated
`labels.untrustedLabelerLogins` config field that doesn't exist — the
real destination is `strip-untrusted-labels.yml`'s
`<labeler-bot-login-N>` placeholder — and a misattributed `ghText`/
`GhTextOptions` reference where the code actually uses `ghApiJson`/
`GhApiJsonOptions`) and one test-coverage gap (the auto-detect path was
untested); all three are fixed in this branch's third commit.

## Linked issue

Closes #2670

## Follow-up issues (if any)

- [x] none

## Background / rationale (if material)

`docs/customization.md`'s Reserved-label guard recipe already documents
the sweep technique in prose (added for #1928); this PR is purely
automation of an already-documented manual procedure, not a new policy.
The helper is deliberately scoped to reporting only — deciding which
candidates are genuinely untrusted and editing the workflow YAML stays
a manual, adopter-owned judgment call, per the issue's acceptance
criteria.

## IDD impact

- [ ] Instruction files changed
- [x] Template files changed
- [x] Helper scripts changed
- [ ] Config schema changed
- [ ] Security / credential / merge behavior changed

## Verification

- [x] `pnpm lint`
- [x] `pnpm test`
- [x] `node scripts/audit-docs.mjs --check`
- [x] `pnpm run docs:sync:check`
- [x] `node scripts/idd-doctor.mjs` (if applicable)

## Safety

- [x] No secret-bearing examples
- [x] No private repo names / URLs
- [x] Merge policy impact reviewed
- [x] Context budget impact reviewed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018jz9naVhMefTusdLL9FzLc

## Summary by CodeRabbit

* **New Features**
  * Added a read-only helper that scans repository issue events to identify and count bot labelers.
  * Supports repository auto-detection, paginated scanning, and table or JSON output.
  * Added safeguards against runaway pagination and mutating command usage.

* **Documentation**
  * Added usage guidance, output details, filtering behavior, and manual review requirements.

* **Tests**
  * Added coverage for filtering, deduplication, pagination, output formats, edge cases, and empty results.

### CODE REVIEW

**Comment by kurone-kito** [comment]
_claude-ecf55bba: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-ecf55bba: critique baseline — IDD automation marker. Do not edit._

File: src/scripts/idd-suggest-untrusted-labelers.mts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+  fetchPage?: (owner: string, repo: string, page: number) => RawIssueEvent[];
+}
+
+/**
+ * Fetch one page of `GET /repos/{owner}/{repo}/issues/events`, projected
+ * down to just `event`/`actor.login`/`actor.type` via a server-side `--jq`
+ * filter (the repository-level endpoint embeds the full parent `issue`
+ * object -- including its body -- in every event, so a page fetched
+ * without this projection, from a repository with thousands of
+ * issues/PRs, can run to several MB).
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-suggest-untrusted-labelers.mts]: **Accepted** — correct catch: `--jq` runs client-side in the `gh` CLI, not server-side. Reworded the doc comment to describe the actual mechanism (gh still receives the full REST response; `--jq` only shrinks what gh writes to this process's own stdout).

File: src/scripts/idd-suggest-untrusted-labelers.mts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+  deps: UntrustedLabelerSweepDeps = {},
+): UntrustedLabelerSweepResult {
+  const fetchPage = deps.fetchPage ?? fetchLabeledBotEventsPage;
+  const events: RawIssueEvent[] = [];
+  let page = 1;
+  let pageCount = 0;
+  for (;;) {
+    const items = fetchPage(owner, repo, page);
+    pageCount += 1;
+    events.push(...items);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-suggest-untrusted-labelers.mts]: **Accepted** — correct: the sweep was accumulating every page's projected events into one array before a single final aggregation pass, which could retain up to ~10,000,000 event objects against the 100,000-page ceiling. Extracted a shared accumulateLabeledBotEvents/finalizeUntrustedLabelerCandidates counting primitive so sweepUntrustedLabelerCandidates now counts each page into a running login->count map as it arrives, bounding memory by the distinct-login count instead of the total event count. aggregateUntrustedLabelerCandidates's public behavior/signature is unchanged.

[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-suggest-untrusted-labelers.mts]: **Accepted** — same finding as the CodeRabbit thread on this same line: the sweep retained every page's projected events in one array instead of aggregating incrementally. Fixed by extracting a shared accumulateLabeledBotEvents/finalizeUntrustedLabelerCandidates counting primitive; sweepUntrustedLabelerCandidates now counts each page into a running login->count map as it arrives and never retains the full event history, bounding memory by the distinct-login count.

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai[bot] did not review HEAD 52c31d560c48ccac5e9c311312e0345720335567 (review limit reached / rate limited); this is not a completed review (source: #issuecomment-5574056498)

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector[bot] did not review HEAD 52c31d560c48ccac5e9c311312e0345720335567 (review still running at the time of this triage pass); this is not a completed review (source: #issuecomment-5574051430)

File: src/scripts/idd-suggest-untrusted-labelers.mts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+  repo: string,
+  deps: UntrustedLabelerSweepDeps = {},
+): UntrustedLabelerSweepResult {
+  const fetchPage = deps.fetchPage ?? fetchLabeledBotEventsPage;
+  const counts = new Map<string, number>();
+  let scannedEventCount = 0;
+  let page = 1;
+  let pageCount = 0;
+  for (;;) {
+    const items = fetchPage(owner, repo, page);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-suggest-untrusted-labelers.mts:311]: **Accepted** — verified: an HTTP 403/429 page-fetch failure (rate limit) previously surfaced as a raw `gh api` stack trace with no indication of the cause. Now caught and re-thrown with an actionable message naming the HTTP status, the page reached, and the remedy (wait for the reset, re-run), via the existing deriveGhHttpStatus helper. Deliberately does NOT add retry/backoff or a resumable cursor -- an exhausted primary rate limit needs a real wait a bounded in-process retry cannot shorten, and no progress is saved between runs either way, so a re-run after the reset simply restarts the sweep. Full rate-limit-aware resumable pagination is a reasonable follow-up for a repository whose full history is large enough to approach the 5,000-requests/hour ceiling; out of scope for this reporting tool today.

File: src/scripts/idd-suggest-untrusted-labelers.mts [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    pageCount += 1;
+    scannedEventCount += items.length;
+    accumulateLabeledBotEvents(items, counts);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-suggest-untrusted-labelers.mts]: **Accepted** — verified empirically against this repository's own live issues/events endpoint: the endpoint sorts newest-first, so a page fetched while new events are still being created can shift an already-fetched event across a page boundary and return it again on a later page. Added the event's own numeric id to the --jq projection and a seenEventIds set tracked across the whole sweep; an id already seen is skipped entirely (not counted again in labeledEventCount or scannedEventCount). Covered by a new test reproducing the exact shifted-window scenario.

**Comment by kurone-kito** [comment]
**Accepted** — chatgpt-codex-connector[bot] confirms its review of current HEAD c5620f5c945e5d1438ff4fc5c37374132cee8930 completed; the findings it raised across this PR are already captured and dispositioned in the resolved review threads above, and this rolling summary comment carries no separate finding of its own.

**Comment by kurone-kito** [comment]
_claude-ecf55bba: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
| Field | Value |
| --- | --- |
| Phase | F3 merged |
| Claim | claude-ecf55bba / claim-8db12f852ee8 |
| Branch | issue/2670-feat-helper-propose-untrusted-labeler |
| Last checked | 2026-09-07T18:28:37.410Z |
| Open blockers | none |
| Next action | F4 cleanup then F5 discover |
| Authoritative by | merge commit 91f929a7c8e7a4582d93450e499fa3e2b7ce96b8; matched head SHA c5620f5c945e5d1438ff4fc5c37374132cee8930 |
